optimade/adapters/structures/__init__.py

```diff
@@ -0,0 +1,89 @@
+# pylint: disable=undefined-variable
+from typing import Union
+
+from optimade.models import StructureResource
+
+from .aiida import get_aiida_structure_data
+from .ase import get_ase_atoms
+from .cif import get_cif
+from .proteindatabank import get_pdb, get_pdbx_mmcif
+from .pymatgen import get_pymatgen_structure
+
+
+__all__ = ("Structure",)
+
+
+class Structure:
+    """Lazy structure converter
+    :param structure: JSON OPTIMADE single structures resource entry.
+    """
+
+    _structure_type_converters = {
+        "aiida_structuredata": get_aiida_structure_data,
+        "ase": get_ase_atoms,
+        "cif": get_cif,
+        "pdb": get_pdb,
+        "pdbx_mmcif": get_pdbx_mmcif,
+        "pymatgen": get_pymatgen_structure,
+    }
+
+    def __init__(self, structure: dict):
+        self._structure = None
+        self._converted_structures = dict.fromkeys(self._structure_type_converters)
+
+        self.structure = structure
+
+    @property
+    def structure(self) -> StructureResource:
+        """Get OPTIMADE structure"""
+        return self._structure
+
+    @structure.setter
+    def structure(self, value: dict):
+        """Set OPTIMADE structure
+        If already set, print that this can _only_ be set once.
+        """
+        if self._structure is None:
+            self._structure = StructureResource(**value)
+        else:
+            print("structure can only be set once and is already set.")
+
+    def __getattr__(self, name: str):
+        """Get converted structure or attribute from OPTIMADE structure
+        Order:
+        - Try to return converted structure if using `get_<valid_structures value>`.
+        - Try to return OPTIMADE StructureResource attribute.
+        - Raise AttributeError
+        """
+        # get_<structure_type>
+        if name.startswith("get_"):
+            structure_type = "_".join(name.split("_")[1:])
+
+            if structure_type not in self._structure_type_converters:
+                raise AttributeError(
+                    f"Non-valid structure type to convert to: {structure_type}. "
+                    f"Valid structure types: {tuple(self._structure_type_converters.keys())}"
+                )
+
+            if self._converted_structures[structure_type] is None:
+                self._converted_structures[
+                    structure_type
+                ] = self._structure_type_converters[structure_type](self.structure)
+
+            return self._converted_structures[structure_type]
+
+        # Try returning StructureResource attribute
+        try:
```

Comment:
I think what you want here is simply
```
return self.structure.__getattr__(name)
```

Reply:
Yes and no.
Yes, because it is indeed what I want, and am getting the same result with `getattr()`.
No, because I also want to display my specialized `AttributeError` message :) In the end, I don't see the difference between `getattr()` and `__getattr__()` in this case - although the latter may be prettier and slightly clearer? Or is there a true benefit to one over the other do you think?

Reply:
Hmm. Thinking more about this, I hesitate. Mainly because we're using internal Python class methods here, and I'd rather go the "proper" way so any `pydantic` logic isn't corrupted or something other - especially thinking of the difference between `__getattribute__()` and `__getattr__()`, where `pydantic` may have implemented something clever for these methods. I could of course just go the Python way and call `__getattribute__()`, but why not just use the more common `getattr()` function?

Reply:
[Here](https://stackoverflow.com/a/1944708/1069467) is a good explanation of the difference.

Now that you're moving out the conversion to `convert`, I'm not sure you still need a custom error message.
The default one should already be quite adequate.

Reply:
> [Here](https://stackoverflow.com/a/1944708/1069467) is a good explanation of the difference.

I looked into the difference as well - which is why I changed my mind, since `pydantic` is surely playing around with these methods. And to be on the safe side, I'd rather go through the proper "Python API" ... :)

> After looking at this, I'm no longer sure what I suggested is indeed the correct solution.
> Your solution with `getattr` might be better.
> 
> However, now that you're moving out the conversion to `convert`, I'm not sure you still need a custom error message.
> The default one should already be quite adequate.

You sure? Having an extra helpful message concerning the available formats and how to get them is not useful?